### PR TITLE
feat(secrets): support first commit results

### PR DESCRIPTION
**Pull request by marynaKK**
## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my feature, policy, or fix is effective and works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/secrets/scan_git_history.py [hunk 1]
```diff
@@ -200,3 +211,50 @@ def _run_scan_one_commit(commit: Commit) -> Tuple[List[RawStore], int]:
                                     rename_from=rename_from, rename_to=rename_to))
             scanned_file_count += 1
         return results, scanned_file_count
+
+    def _scan_first_commit(self):
```
Comment (omryMen) [inline checkov/secrets/scan_git_history.py:215]: why not first insert it to the commits_diffs ?

File: checkov/secrets/git_types.py [hunk 2]
```diff
@@ -25,17 +25,19 @@ class RenamedFile(TypedDict):
...
             metadata: CommitMetadata,
             files: dict[str, CommitDiff] | None = None,
-            renamed_files: dict[str, RenamedFile] | None = None
+            renamed_files: dict[str, RenamedFile] | None = None,
+            is_first: bool = False
     ):
         self.metadata: CommitMetadata = metadata
         self.files: dict[str, CommitDiff] = files or {}
         self.renamed_files: dict[str, RenamedFile] = renamed_files or {}
+        self.is_first = is_first
```
Comment (omryMen) [inline checkov/secrets/git_types.py:40]: is first can be inside metadata

File: checkov/secrets/scan_git_history.py [hunk 3]
```diff
@@ -200,3 +211,50 @@ def _run_scan_one_commit(commit: Commit) -> Tuple[List[RawStore], int]:
...
+                    potential_secret.is_added = True
+                logging.info(
+                    f"Found {len(file_results)} secrets in file path {file_name} in first commit {first_commit_sha}")
+
+                commit_diff.add_file(filename=file_path, commit_diff=file_content)
+                self.raw_store.append(RawStore(file_results=file_results, file_name=file_path, commit=commit_diff,
+                                               type=FILE_RESULTS_STR, rename_from='', rename_to=''))
+
+        if os.path.exists(tmp_git_history_dir):
+            shutil.rmtree(tmp_git_history_dir)
```
Comment (omryMen) [inline checkov/secrets/scan_git_history.py:260]: you can just do rmtree with ignore_errors=True

File: checkov/secrets/scan_git_history.py [hunk 4]
```diff
@@ -39,13 +42,16 @@ def __init__(self, root_folder: str, secrets: SecretsCollection,
...
         self.raw_store: List[RawStore] = []
-        self.commits_diff: List[Commit] = []
+        self.repo: Repo
 
     def scan_history(self, last_commit_scanned: Optional[str] = '') -> bool:
         """return true if the scan finished without timeout"""
         timeout_class = ThreadingTimeout if platform.system() == 'Windows' else SignalTimeout
         # mark the scan to finish within the timeout
         with timeout_class(self.timeout) as to_ctx_mgr:
+            if not self.set_repo():
```
Comment (omryMen) [inline checkov/secrets/scan_git_history.py:52]: why not in init ?

Comment (marynaKK) [reply checkov/secrets/scan_git_history.py:52]: tests :( other ides maybe to get it work with tests?

File: checkov/secrets/scan_git_history.py [hunk 5]
```diff
@@ -92,27 +98,33 @@ def _create_secret_collection(self) -> None:
...
         logging.info(f"Created secret collection for {len(self.history_store.secrets_by_file_value_type)} secrets")
 
+    def set_repo(self, root_folder: str | None = None) -> bool:
+        if not root_folder:
+            root_folder = self.root_folder
+        if git_import_error is not None:
+            logging.warning(f"Unable to load git module (is the git executable available?) {git_import_error}")
+            return False
+        try:
+            self.repo = git.Repo(root_folder)
```
Comment (nimrodkor) [inline checkov/secrets/scan_git_history.py:108]:
```suggestion
            self.repo = Repo(root_folder)
```

**Review by nimrodkor — COMMENTED** [review]
Very nice

File: checkov/secrets/git_history_store.py [hunk 6]
```diff
@@ -60,7 +60,7 @@ def _add_new_secret(self, secret_key: str, secret: PotentialSecret, commit: Comm
                                                                        'removed_commit_hash': '',
                                                                        'removed_date': ''})
                 return
-        code_line = search_for_code_line(commit.files.get(secret.filename, ''), secret.secret_value, secret.is_added)
+        code_line = search_for_code_line(commit.files[secret.filename], secret.secret_value, secret.is_added)
```
[the remark this answers is not in the document]
Comment (marynaKK) [reply checkov/secrets/git_history_store.py:63]: cause we don't get the error anymore, didn't want to mask it with a get.

File: checkov/secrets/scan_git_history.py [hunk 7]
```diff
@@ -92,27 +95,32 @@ def _create_secret_collection(self) -> None:
...
+            logging.warning(f"Unable to load git module (is the git executable available?) {git_import_error}")
+            return False
+        try:
+            self.repo = Repo(root_folder)
+            return True
+        except Exception as e:
+            logging.error(f"Folder {root_folder} is not a GIT project {e}")
+            return False
+
     def _get_commits_diff(self, root_folder: str, last_commit_sha: Optional[str] = None) -> List[Commit]:
```
Comment (omryMen) [inline checkov/secrets/scan_git_history.py:111]: sounds like we no longer need root folder in this function since it has it in self

File: checkov/secrets/scan_git_history.py (hunk 4 above)
Comment (omryMen) [reply checkov/secrets/scan_git_history.py:50]: other idea is to inject it between functions. not sure what is better

File: checkov/secrets/scan_git_history.py [hunk 8]
```diff
@@ -54,17 +64,16 @@ def scan_history(self, last_commit_scanned: Optional[str] = '') -> bool:
...
 
     def _scan_history(self, last_commit_scanned: Optional[str] = '') -> bool:
-        # commits_diff = self._get_commits_diff(self.root_folder, last_commit_sha=last_commit_scanned)
-        self.commits_diff = self._get_commits_diff(self.root_folder, last_commit_sha=last_commit_scanned)
-        if not self.commits_diff:
-            return False
-        logging.info(f"[_scan_history] got {len(self.commits_diff)} files diffs in {self.commits_count} commits")
+        commits_diff: List[Commit] = []
+        commits_diff.extend(self._get_first_commit())
+        commits_diff.extend(self._get_commits_diff(last_commit_sha=last_commit_scanned))
```
Comment (nimrodkor) [inline checkov/secrets/scan_git_history.py:69]: a more pythonic way to generate this:
```suggestion
        commits_diff: List[Commit] = [*self._get_first_commit(), *self._get_commits_diff(last_commit_sha=last_commit_scanned)]
```

File: checkov/secrets/scan_git_history.py [hunk 9]
```diff
@@ -54,17 +64,16 @@ def scan_history(self, last_commit_scanned: Optional[str] = '') -> bool:
         return scanned
 
     def _scan_history(self, last_commit_scanned: Optional[str] = '') -> bool:
-        # commits_diff = self._get_commits_diff(self.root_folder, last_commit_sha=last_commit_scanned)
-        self.commits_diff = self._get_commits_diff(self.root_folder, last_commit_sha=last_commit_scanned)
-        if not self.commits_diff:
-            return False
-        logging.info(f"[_scan_history] got {len(self.commits_diff)} files diffs in {self.commits_count} commits")
+        commits_diff: List[Commit] = []
+        commits_diff.extend(self._get_first_commit())
```
Comment (omryMen) [inline checkov/secrets/scan_git_history.py:68]: I'm not sure if we need it in the case when we have the last_commit_scanned

**Review by omryMen — APPROVED** [review]
just one comment to check.. overall looks great

File: checkov/secrets/scan_git_history.py [hunk 10]
```diff
@@ -200,3 +214,29 @@ def _run_scan_one_commit(commit: Commit) -> Tuple[List[RawStore], int]:
                                     rename_from=rename_from, rename_to=rename_to))
             scanned_file_count += 1
         return results, scanned_file_count
+
+    @time_it
+    def _get_first_commit(self) -> List[Commit]:
```
Comment (nimrodkor) [inline checkov/secrets/scan_git_history.py:219]: @marynaKK I'm surprised this returns a list and not just a commit... The first one

**Review by nimrodkor — COMMENTED** [review]
Just a little readability comments

**Review by lirshindalman — APPROVED** [review]
:rocket:

File: checkov/secrets/scan_git_history.py (hunk 10 above)
Comment (marynaKK) [reply checkov/secrets/scan_git_history.py:219]: you are right. I need to adjust the test to return an empty commit.

File: checkov/secrets/scan_git_history.py (hunk 9 above)
Comment (marynaKK) [reply checkov/secrets/scan_git_history.py:68]: ohhhhhh this is good! you are right!!!!! thank you!
